geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureConverterFactory.scala

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.locationtech.geomesa.convert
+
+
+import javax.imageio.spi.ServiceRegistry
+
+import com.typesafe.config.Config
+import com.typesafe.scalalogging.slf4j.Logging
+import org.locationtech.geomesa.convert.Transformers.{EvaluationContext, Expr}
+import org.locationtech.geomesa.feature.AvroSimpleFeatureFactory
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.JavaConversions._
+import scala.util.{Failure, Success, Try}
+
+trait Field {
+  def name: String
+  def transform: Transformers.Expr
+  def eval(args: Any*)(implicit ec: EvaluationContext): Any = transform.eval(args: _*)
+}
+
+case class SimpleField(name: String, transform: Transformers.Expr) extends Field
+
+trait SimpleFeatureConverterFactory[I] {
+
+  def canProcess(conf: Config): Boolean
+
+  def canProcessType(conf: Config, name: String) = Try { conf.getString("type").equals(name) }.getOrElse(false)
+
+  def buildConverter(sft: SimpleFeatureType, conf: Config): SimpleFeatureConverter[I]
+
+  def buildFields(fields: Seq[Config]): IndexedSeq[Field] =
+    fields.map { f =>
+      val name = f.getString("name")
+      val transform = Transformers.parseTransform(f.getString("transform"))
+      SimpleField(name, transform)
+    }.toIndexedSeq
+
+  def buildIdBuilder(t: String) = Transformers.parseTransform(t)
+
+}
+
+object SimpleFeatureConverters {
+  val providers = ServiceRegistry.lookupProviders(classOf[SimpleFeatureConverterFactory[_]]).toList
+
+  def build[I](sft: SimpleFeatureType, conf: Config) = {
+    val converterConfig = conf.getConfig("converter")
+    providers
+      .find(_.canProcess(converterConfig))
+      .map(_.buildConverter(sft, converterConfig).asInstanceOf[SimpleFeatureConverter[I]])
+      .getOrElse(throw new IllegalArgumentException("Cannot find factory"))
+  }
+}
+
+trait SimpleFeatureConverter[I] {
+  def targetSFT: SimpleFeatureType
+  def processInput(is: Iterator[I]): Iterator[SimpleFeature]
+  def processSingleInput(i: I): Option[SimpleFeature]
+  def close(): Unit = {}
+}
+
+trait ToSimpleFeatureConverter[I] extends SimpleFeatureConverter[I] with Logging {
+  def logErrors: Boolean = false
+  def targetSFT: SimpleFeatureType
+  def inputFields: IndexedSeq[Field]
+  def idBuilder: Expr
+  def fromInputType(i: I): Array[Any]
+  val fieldNameMap = inputFields.zipWithIndex.map { case (f, idx) => (f.name, idx)}.toMap
+
+  val featureFactory = new AvroSimpleFeatureFactory
+
+  implicit val ctx = new EvaluationContext(fieldNameMap, null)
+
+  def convert(t: Array[Any], reuse: Array[Any], sfAttrReuse: Array[Any]): SimpleFeature =
+  {
+    val attributes =
+      if(reuse == null) Array.ofDim[Any](inputFields.length)
+      else reuse
+    ctx.computedFields = attributes
+
+    inputFields.zipWithIndex.foreach { case (field, i) =>
```

Comment:
Is this function going to in inner loops/called alot?  If so, I'd favor a while loop over zipWithIndex.foreach.